lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -1040,21 +1084,27 @@ swift::castValueToABICompatibleType(SILBuilder *builder, SILLocation loc,
         builder->createOptionalSome(loc, value, loweredOptionalSrcType);
     // Cast the wrapped value.
     return castValueToABICompatibleType(builder, loc, wrappedValue,
-                                        wrappedValue->getType(), destTy);
+                                        wrappedValue->getType(), destTy,
+                                        usePoints);
   }
 
   // Handle tuple types.
   // Extract elements, cast each of them, create a new tuple.
   if (auto srcTupleTy = srcTy.getAs<TupleType>()) {
     SmallVector<SILValue, 8> expectedTuple;
     bool changedCFG = false;
+    DestructureTupleInst *dsi = nullptr;
+    if (value.getOwnershipKind() == OwnershipKind::Owned) {
+      dsi = builder->createDestructureTuple(loc, value);
+    }
     for (unsigned i = 0, e = srcTupleTy->getNumElements(); i < e; ++i) {
-      SILValue element = builder->createTupleExtract(loc, value, i);
+      SILValue element =
+        dsi ? dsi->getResult(i) : builder->createTupleExtract(loc, value, i);
```

Comment:
There is a helper on SILBuilder that does this. (emitDestructureOperation I think?)

Reply:
Yes, that works. It also forces the destructure for guaranteed values, which is good. We don't want struct/tuple_extract in OSSA if we can avoid it.